src/network/message.rs

```diff
@@ -37,21 +37,64 @@ use consensus::encode::MAX_VEC_SIZE;
 #[derive(PartialEq, Eq, Clone, Debug)]
 pub struct CommandString(Cow<'static, str>);
 
+impl CommandString {
+    /// Convert from various string types into a [CommandString].
+    ///
+    /// Supported types are:
+    /// - `&'static str`
+    /// - `String`
+    /// - `&str`
+    ///
+    /// Returns an empty error if and only if the string is
+    /// larger than 12 characters in length.
+    pub fn from<S: Into<Cow<'static, str>>>(s: S) -> Result<CommandString, ()> {
+        let cow = s.into();
+        if cow.as_ref().len() > 12 {
+            Err(())
+        } else {
+            Ok(CommandString(cow))
+        }
+    }
+}
+
 impl fmt::Display for CommandString {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         f.write_str(self.0.as_ref())
     }
 }
 
+impl str::FromStr for CommandString {
+    type Err = ();
+    fn from_str(s: &str) -> Result<CommandString, ()> {
+        // Don't use CommandString::from to avoid allocating when
+        // it's not needed.
```

Comment:
You're saying don't call `CommandString::from` to *avoid allocating* but you then call `to_owned()` which allocates?

Reply:
Hmm, I think I might have meant `Use` instead of `Don't use`.